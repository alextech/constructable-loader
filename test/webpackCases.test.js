import fs from "fs";

const webpack = require('webpack');
const path = require('path');
const rimraf = require('rimraf');

function runWebpack(options/*: Configuration*/) /*: Promise<webpack.Stats>*/ {
    const compiler = webpack(options);
    return new Promise /* <webpack.Stats>*/((resolve, reject) => {
        compiler.run((err /*: Error*/, stats /*: Stats*/) => {
            if (err) reject(err);
            if (stats.hasErrors()) reject(new Error(stats.toString()));
            resolve(stats);
        });
    });
}

async function readFileOrEmpty(path/*: string*/) /*: Promise<string>*/ {
    try {
        return await fs.promises.readFile(path, "utf-8");
        // eslint-disable-next-line no-empty
    } catch (e) {
        return "";
    }
}

const cases = [
    'simple',
    'simple-with-purge',
    'multiple-with-postcss-loader',
    'edgecase-comment-has-backtick'
];

describe("Webpack integration", () => {
    for (const testCase of cases) {
        it(`works with ${testCase} configuration`, async () => {
            const outputDirectory = path.resolve(__dirname, 'webpack_bundle', testCase);
            rimraf.sync(outputDirectory);

            const testDirectory = path.resolve(__dirname, 'cases', testCase);

            process.chdir(testDirectory);

            const webpackConfig = require(`${testDirectory}/webpack.config.js`);

            await runWebpack({
               ...webpackConfig,
               output: {
                   path: outputDirectory,
               }
            });

            let webpackModule = (await readFileOrEmpty(path.resolve(outputDirectory, 'unit_test.js')))
                .replace(/(?:\\[rn])+/g, "\n");

            expect(webpackModule).toMatchSnapshot(testCase);
        });
    }
});

