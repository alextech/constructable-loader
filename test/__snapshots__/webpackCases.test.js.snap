// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webpack integration works with edgecase-comment-has-backtick configuration: edgecase-comment-has-backtick 1`] = `
"(window[\\"webpackJsonp\\"] = window[\\"webpackJsonp\\"] || []).push([[\\"unit_test\\"],{

/***/ \\"./src/index.js\\":
/*!**********************!*\\\\
  !*** ./src/index.js ***!
  \\\\**********************/
/*! no exports provided */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
eval(\\"__webpack_require__.r(__webpack_exports__);
/* harmony import */ var _panel_js__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./panel.js */ \\\\\\"./src/panel.js\\\\\\");
//# sourceURL=webpack:///./src/index.js?\\");

/***/ }),

/***/ \\"./src/panel.js\\":
/*!**********************!*\\\\
  !*** ./src/panel.js ***!
  \\\\**********************/
/*! no exports provided */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
eval(\\"__webpack_require__.r(__webpack_exports__);
/* harmony import */ var _style_css__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./style.css */ \\\\\\"./src/style.css\\\\\\");
class Panel extends HTMLElement {
    constructor() {
        super();
        this.attachShadow({mode: \\\\\\"open\\\\\\"});
        this.shadowRoot.innerHTML = \`<div class=\\\\\\"panel\\\\\\">Panel CMP</div>\`;
        this.shadowRoot.adoptedStyleSheets = [_style_css__WEBPACK_IMPORTED_MODULE_0__[\\\\\\"default\\\\\\"]]
    }
}
if (customElements.get('ata-panel') == null) {
    customElements.define('ata-panel', Panel);
}
//# sourceURL=webpack:///./src/panel.js?\\");

/***/ }),

/***/ \\"./src/style.css\\":
/*!***********************!*\\\\
  !*** ./src/style.css ***!
  \\\\***********************/
/*! exports provided: default */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
eval(\\"__webpack_require__.r(__webpack_exports__);
const styleSheet = new CSSStyleSheet(); styleSheet.replaceSync(\`.btn {
    background-color: blue;
}
.panel {
    background-color: orange;
    border-style: solid;
}
.white-listed {
    background-color: white;
}
/*
 * Discovered when adding tailwind
 */
/**
 * Render the \\\\\\\\\`main\\\\\\\\\` element consistently in IE.
 */
\`); /* harmony default export */ __webpack_exports__[\\\\\\"default\\\\\\"] = (styleSheet);
//# sourceURL=webpack:///./src/style.css?\\");

/***/ })

},[[\\"./src/index.js\\",\\"runtime\\"]]]);"
`;

exports[`Webpack integration works with multiple-with-postcss-loader configuration: multiple-with-postcss-loader 1`] = `
"(window[\\"webpackJsonp\\"] = window[\\"webpackJsonp\\"] || []).push([[\\"unit_test\\"],{

/***/ \\"./src/form.js\\":
/*!*********************!*\\\\
  !*** ./src/form.js ***!
  \\\\*********************/
/*! no exports provided */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
eval(\\"__webpack_require__.r(__webpack_exports__);
/* harmony import */ var _style_css__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./style.css */ \\\\\\"./src/style.css\\\\\\");
class Form extends HTMLElement {
    constructor() {
        super();
        this.attachShadow({mode: \\\\\\"open\\\\\\"});
        this.shadowRoot.innerHTML = \`<div class=\\\\\\"form\\\\\\"><form>Form CMP</form></div>\`;
        this.shadowRoot.adoptedStyleSheets = [_style_css__WEBPACK_IMPORTED_MODULE_0__[\\\\\\"default\\\\\\"]]
    }
}
if (customElements.get('ata-panel') == null) {
    customElements.define('ata-panel', Form);
}
//# sourceURL=webpack:///./src/form.js?\\");

/***/ }),

/***/ \\"./src/index.js\\":
/*!**********************!*\\\\
  !*** ./src/index.js ***!
  \\\\**********************/
/*! no exports provided */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
eval(\\"__webpack_require__.r(__webpack_exports__);
/* harmony import */ var _panel__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./panel */ \\\\\\"./src/panel.js\\\\\\");
/* harmony import */ var _form__WEBPACK_IMPORTED_MODULE_1__ = __webpack_require__(/*! ./form */ \\\\\\"./src/form.js\\\\\\");
//# sourceURL=webpack:///./src/index.js?\\");

/***/ }),

/***/ \\"./src/panel.js\\":
/*!**********************!*\\\\
  !*** ./src/panel.js ***!
  \\\\**********************/
/*! no exports provided */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
eval(\\"__webpack_require__.r(__webpack_exports__);
/* harmony import */ var _style_css__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./style.css */ \\\\\\"./src/style.css\\\\\\");
class Panel extends HTMLElement {
    constructor() {
        super();
        this.attachShadow({mode: \\\\\\"open\\\\\\"});
        this.shadowRoot.innerHTML = \`<div class=\\\\\\"panel\\\\\\">Panel CMP</div>\`;
        this.shadowRoot.adoptedStyleSheets = [_style_css__WEBPACK_IMPORTED_MODULE_0__[\\\\\\"default\\\\\\"]]
    }
}
if (customElements.get('ata-form') == null) {
    customElements.define('ata-form', Panel);
}
//# sourceURL=webpack:///./src/panel.js?\\");

/***/ }),

/***/ \\"./src/style.css\\":
/*!***********************!*\\\\
  !*** ./src/style.css ***!
  \\\\***********************/
/*! exports provided: default */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
eval(\\"__webpack_require__.r(__webpack_exports__);
const styleSheet = new CSSStyleSheet(); styleSheet.replaceSync(\`.panel {
    background-color: orange;
    border-style: solid;
}
.white-listed {
    background-color: white;
}
form {
    padding-left: 1em;
}
\`); /* harmony default export */ __webpack_exports__[\\\\\\"default\\\\\\"] = (styleSheet);
//# sourceURL=webpack:///./src/style.css?\\");

/***/ })

},[[\\"./src/index.js\\",\\"runtime\\"]]]);"
`;

exports[`Webpack integration works with simple configuration: simple 1`] = `
"(window[\\"webpackJsonp\\"] = window[\\"webpackJsonp\\"] || []).push([[\\"unit_test\\"],{

/***/ \\"./src/index.js\\":
/*!**********************!*\\\\
  !*** ./src/index.js ***!
  \\\\**********************/
/*! no exports provided */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
eval(\\"__webpack_require__.r(__webpack_exports__);
/* harmony import */ var _panel_js__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./panel.js */ \\\\\\"./src/panel.js\\\\\\");
//# sourceURL=webpack:///./src/index.js?\\");

/***/ }),

/***/ \\"./src/panel.js\\":
/*!**********************!*\\\\
  !*** ./src/panel.js ***!
  \\\\**********************/
/*! no exports provided */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
eval(\\"__webpack_require__.r(__webpack_exports__);
/* harmony import */ var _style_css__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./style.css */ \\\\\\"./src/style.css\\\\\\");
class Panel extends HTMLElement {
    constructor() {
        super();
        this.attachShadow({mode: \\\\\\"open\\\\\\"});
        this.shadowRoot.innerHTML = \`<div class=\\\\\\"panel\\\\\\">Panel CMP</div>\`;
        this.shadowRoot.adoptedStyleSheets = [_style_css__WEBPACK_IMPORTED_MODULE_0__[\\\\\\"default\\\\\\"]]
    }
}
if (customElements.get('ata-panel') == null) {
    customElements.define('ata-panel', Panel);
}
//# sourceURL=webpack:///./src/panel.js?\\");

/***/ }),

/***/ \\"./src/style.css\\":
/*!***********************!*\\\\
  !*** ./src/style.css ***!
  \\\\***********************/
/*! exports provided: default */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
eval(\\"__webpack_require__.r(__webpack_exports__);
const styleSheet = new CSSStyleSheet(); styleSheet.replaceSync(\`.btn {
    background-color: blue;
}
.panel {
    background-color: orange;
    border-style: solid;
}
.white-listed {
    background-color: white;
}
.\\\\\\\\\\\\\\\\32xl\\\\\\\\:ring-offset-white {
    --tw-ring-offset-color: #fff;
}
\`); /* harmony default export */ __webpack_exports__[\\\\\\"default\\\\\\"] = (styleSheet);
//# sourceURL=webpack:///./src/style.css?\\");

/***/ })

},[[\\"./src/index.js\\",\\"runtime\\"]]]);"
`;

exports[`Webpack integration works with simple-with-purge configuration: simple-with-purge 1`] = `
"(window[\\"webpackJsonp\\"] = window[\\"webpackJsonp\\"] || []).push([[\\"unit_test\\"],{

/***/ \\"./src/index.js\\":
/*!**********************!*\\\\
  !*** ./src/index.js ***!
  \\\\**********************/
/*! no exports provided */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
eval(\\"__webpack_require__.r(__webpack_exports__);
/* harmony import */ var _panel_js__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./panel.js */ \\\\\\"./src/panel.js\\\\\\");
//# sourceURL=webpack:///./src/index.js?\\");

/***/ }),

/***/ \\"./src/panel.js\\":
/*!**********************!*\\\\
  !*** ./src/panel.js ***!
  \\\\**********************/
/*! no exports provided */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
eval(\\"__webpack_require__.r(__webpack_exports__);
/* harmony import */ var _style_css__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./style.css */ \\\\\\"./src/style.css\\\\\\");
class Panel extends HTMLElement {
    constructor() {
        super();
        this.attachShadow({mode: \\\\\\"open\\\\\\"});
        this.shadowRoot.innerHTML = \`<div class=\\\\\\"panel\\\\\\">Panel CMP</div>\`;
        this.shadowRoot.adoptedStyleSheets = [_style_css__WEBPACK_IMPORTED_MODULE_0__[\\\\\\"default\\\\\\"]]
    }
}
if (customElements.get('ata-panel') == null) {
    customElements.define('ata-panel', Panel);
}
//# sourceURL=webpack:///./src/panel.js?\\");

/***/ }),

/***/ \\"./src/style.css\\":
/*!***********************!*\\\\
  !*** ./src/style.css ***!
  \\\\***********************/
/*! exports provided: default */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
eval(\\"__webpack_require__.r(__webpack_exports__);
const styleSheet = new CSSStyleSheet(); styleSheet.replaceSync(\`.white-listed {
    background-color: white;
}
\`); /* harmony default export */ __webpack_exports__[\\\\\\"default\\\\\\"] = (styleSheet);
//# sourceURL=webpack:///./src/style.css?\\");

/***/ })

},[[\\"./src/index.js\\",\\"runtime\\"]]]);"
`;
